client/recordUtil.js

```diff
@@ -0,0 +1,131 @@
+'use strict'
+
+const proto = require('./constants/proto')
+
+module.exports = class RecordUtil {
+  /**
+   * @param {Serializer} serializer
+   */
+  constructor (serializer) {
+    if (!serializer) { throw new Error('Missing serializer.') }
+    this.serializer = serializer
+  }
+
+  get SyncRecord () {
+    return this.serializer.api.SyncRecord
+  }
+
+  get Bookmark () {
+    return this.SyncRecord.lookup('Bookmark')
+  }
+
+  get Device () {
+    return this.SyncRecord.lookup('Device')
+  }
+
+  get Site () {
+    return this.SyncRecord.lookup('Site')
+  }
+
+  get SiteSetting () {
+    return this.SyncRecord.lookup('SiteSetting')
+  }
+
+  /**
+   * In case of UPDATEs to nonexistant records, try falling back to CREATE.
+   * @param {SyncRecord} record
+   * @returns {SyncRecord=} SyncRecord with action CREATE; or null.
+   */
+  static createFromUpdate (record) {
+    if (!record || record.action !== proto.actions.UPDATE) {
+      throw new Error('Missing UPDATE syncRecord.')
+    }
+    switch (record.objectData) {
+      case 'bookmark':
+        return this.createFromUpdateBookmark(record)
+      case 'historySite':
+        return this.createFromUpdateHistorySite(record)
+      case 'siteSetting':
+        return this.createFromUpdateSiteSetting(record)
+      default:
+        return null
+    }
+  }
+
+  static createFromUpdateBookmark (record) {
+    return null
+    // if (!record.historySite || !record.historySite.location) {
+    //   return null
+    // }
+    // const SyncRecord = record.constructor
+  }
+
+  static createFromUpdateHistorySite (record) {
+    if (!record.historySite || !record.historySite.location) {
+      return null
+    }
+    const site = record.historySite
+    const newSite = {
+      location: site.location,
+      title: site.title || site.location,
+      customTitle: site.customTitle || undefined,
+      lastAccessedTime: site.lastAccessedTime || Date.now(),
+      creationTime: site.creationTime || Date.now()
+    }
+    return record.$type.create(Object.assign(
```

Comment:
to confirm: a protobuf object is created here but not encoded into bytes, so if it's sent to the browser via IPC will it be able to access the object?